Make nltk.wsd.lesk() deterministic.

dimazest opened the pull request.

dimazest committed:
Skip nltk.wsd docktest because lesk() return value is nondeterministic.

dimazest committed:
wsd.doctest: skip lesk() test.

stevenbird commented:
I think it's a bug if the Lesk algorithm is not deterministic. We've had to do some work on other WordNet methods to ensure that they are deterministic (e.g. with sorting results).

dimazest commented:
Ok. I'll make it deterministic.

dimazest committed:
Merge branch 'develop' into wsd-tests

dimazest committed:
nltk.wds.lesk() is deterministic now.

There is no need for the dictionary lesk() argument to be a dictionary.
Now, it’s an iterable of sorted items.

dimazest commented on nltk/wsd.py:
```diff
@@ -12,53 +12,61 @@
     :param ambiguous_word: The ambiguous word that requires WSD.
     :param pos: A specified Part-of-Speech (POS).
-    :param dictionary: A list of words that 'signifies' the ambiguous word.
+    :param dict dictionary: A mapping of synsets to their definitions.
```
I'm not sure whether it makes sense for dictionary to be of type dict, but i've left it to keep compatibility.

stevenbird replied:
I'd prefer to break compatibility in favour of correctness, since NLTK 3 is not widely adopted yet. So what about a better choice of name and type for this parameter?

dimazest replied:
i'll have a look to the paper to see how things are named there.

stevenbird replied:
OK, thanks. Similarity with the published algorithm helps us read and maintain the code.

dimazest replied:
I agree. Keeping code as close to a paper could be a good NLTK coding guide. Also, implementing examples in a paper as tests might be a good idea.

dimazest commented:
Hi,

I've cleaned up lesk() implementation, so all the tests pass now!!!

dimazest committed:
Merge with development

dimazest committed:
Refactoring of wsd.lesk().

dimazest committed:
Whitespaces

dimazest commented on nltk/wsd.py:
```diff
@@ -6,59 +6,41 @@
+
+    _, sense = max(
+        (len(context.intersection(ss.definition().split())), ss)
+        for ss in synsets if pos is None or str(ss.pos()) == pos
```
I use `if pos is None *or* ...` so when no pos tag is passed all possible senses are evaluated.
